moveit_core/kinematics_base/CMakeLists.txt

```diff
@@ -1,18 +1,38 @@
+cmake_minimum_required(VERSION 3.5)
 set(MOVEIT_LIB_NAME moveit_kinematics_base)
 
+# Default to C++14
+if(NOT CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 14)
+endif()
+
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
 add_library(${MOVEIT_LIB_NAME} src/kinematics_base.cpp)
-set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES VERSION "${${PROJECT_NAME}_VERSION}")
+
+# set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES VERSION ${${PROJECT_NAME}_VERSION})
+
 # Avoid warnings about deprecated members of KinematicsBase when building KinematicsBase itself.
 # TODO: remove when deperecated variables (tip_frame_, search_discretization_) are finally removed.
-target_compile_options(${MOVEIT_LIB_NAME} PRIVATE -Wno-deprecated-declarations)
+# target_compile_options(${MOVEIT_LIB_NAME} PRIVATE -Wno-deprecated-declarations)
 
 # This line is needed to ensure that messages are done being built before this is built
-add_dependencies(${MOVEIT_LIB_NAME} ${catkin_EXPORTED_TARGETS})
+# add_dependencies(${MOVEIT_LIB_NAME} ${catkin_EXPORTED_TARGETS})
+
+
+target_link_libraries(${MOVEIT_LIB_NAME} ${rclcpp_LIBRARIES} ${rmw_implementation_LIBRARIES} ${urdfdom_LIBRARIES} ${urdfdom_headers_LIBRARIES} ${Boost_LIBRARIES} ${common_interfaces_LIBRARIES})
+ament_target_dependencies(${MOVEIT_LIB_NAME}
+  urdf
+  urdfdom_headers)
```

Comment:
@mlautman ?

Reply:
Ament/Colcon is A LOT more free-form than Catkin which is very frustrating from the perspective of new users. I am not very knowledgeable about cmake but this would be a meaningful conversation to have with @dirk-thomas

Reply:
You don't need `target_link_libraries` _and_ `ament_target_dependencies`. For each dependency given to `ament_target_dependencies` it will (attempt) to extract the libraries and other settings and add them to the target given, see:

https://github.com/ament/ament_cmake/blob/409ee0403d8d1eba0dc5a58cf52dcec5eadc91d2/ament_cmake_target_dependencies/cmake/ament_target_dependencies.cmake#L15-L30

> Ament/Colcon is A LOT more free-form than Catkin

What does free-form mean exactly?

If you guys want to type less lines, at the expense of precision, you can try `ament_cmake_auto`:

https://github.com/ament/ament_cmake/tree/master/ament_cmake_auto

Unfortunately we haven't had time to add examples and details documentation, but it will do things like automatically add the include dirs and link libraries to your topics based on the dependencies in your `package.xml`.

I had a prototype of something I called `ros_cmake_auto` (`ament_cmake_auto` + ROS specific stuff) that used it (indirectly) here:

https://github.com/wjwwood/ros_cmake_auto_examples

Example:

https://github.com/wjwwood/ros_cmake_auto_examples/blob/master/cpp_examples/node_with_publisher/CMakeLists.txt

And using `ament_cmake_auto` here: 

- https://github.com/wjwwood/ros_cmake_auto_examples/blob/532f56417db60052cfb46174cae8cbfeae6105a2/ros_cmake_auto/cmake/ros_cmake_auto_add_node.cmake#L28-L31
- https://github.com/wjwwood/ros_cmake_auto_examples/blob/532f56417db60052cfb46174cae8cbfeae6105a2/ros_cmake_auto/cmake/ros_cmake_auto_end.cmake#L4

Reply:
The [docblock of `ament_target_dependencies`](https://github.com/ament/ament_cmake/blob/207a07b72af1e5a780887d7c2268df57ef456af3/ament_cmake_target_dependencies/cmake/ament_target_dependencies.cmake#L15-L30) describes what it is doing. It simply provides a convenience function so that you don't have to call `target_include_directories`, `target_link_libraries`, etc. separately. It only works when the dependency defines standardized variable names (e.g. `xxx_INCLUDE_DIRS`, `xxx_LIBRARIES`).

So calling `target_link_libraries` and `ament_target_dependencies` for the same dependency is redundant.

Reply:
I don't really recommend the simplified interfaces for a core library like moveit, however. I think it should be precise and careful about what it links against what and what it exports to other packages.

I'd recommend learning cmake and what all the parts do (even if you find it frustrating) because they have a purpose.

Reply:
@wjwwood @dirk-thomas I appreciate all of the additional information. I understand that these tools and documentation are still a work in progress. By "free-form" I only mean that as of yet best practices have not yet emerged and finding examples to use as boilerplate can be difficult. 

<quick_rant>
One of ROS1's strengths is that catkin makes it very simple to create a ROS1 application. The `catkin_create_pkg` tool provides users with a heavily commented package that works out of the box and establishes best practices that covers 99% of use cases. By contrast, `ros2 pkg create` produces a bare-bones package with minimal documentation that doesn't really do much. I fully understand that time is limited but speaking as a ROS1 developer who is migrating to ROS2, I can't overstate how important documentation and best practices are to driving adoption and community building.
</quick_rant>

Reply:
As always this boils down to resources. Someone needs to spend a significant amount of time and resources to write good tutorials as we are used to from ROS 1. And since we are fairly resource constraint we can't do everything which would be nice to have/do at the same time. As a consequence we have been focusing our limited resources to develop the necessary features and functionality so that ROS 2 provides all the necessary foundations available in ROS 1 (pub/sub, services, actions, C++, Python, command line tool, rviz, rqt, components, etc.). At some point we will hopefully be able to spend more effort on documentation - it is just something nobody considers paying for :wink:

Reply:
I think part of the thinking was also that "you can do so many things, that it would be cluttered to put them all in one template file", so there is also this repository which has full package examples for many different cases, but it too needs documentation and more examples:

https://github.com/ros2/examples/tree/master/rclcpp

The above is sort of like what other projects call a "cookbook".

There's also a repository for tutorials, which are intended to be more heavily documented (commented) and less varied, but geared more for teaching basics:

https://github.com/ros2/tutorials

But they are sadly out of date and we don't even test them regularly (we build the aforementioned examples on every new pr however).

As dirk said, we have plenty of ideas how to improve on-boarding, but it's always behind feature development in priorities.